{
 "cells": [
  {
   "cell_type": "code",
   "execution_count": 1,
   "id": "164f305a-4b56-4e36-b267-9066b75c748b",
   "metadata": {},
   "outputs": [],
   "source": [
    "import pandas as pd"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 2,
   "id": "668c6b28-04d7-48a8-993c-b34d5e16bded",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>0</th>\n",
       "      <th>1</th>\n",
       "      <th>2</th>\n",
       "      <th>3</th>\n",
       "      <th>4</th>\n",
       "      <th>5</th>\n",
       "      <th>6</th>\n",
       "      <th>7</th>\n",
       "      <th>8</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>chr1</td>\n",
       "      <td>HAVANA</td>\n",
       "      <td>gene</td>\n",
       "      <td>3143476</td>\n",
       "      <td>3144545</td>\n",
       "      <td>.</td>\n",
       "      <td>+</td>\n",
       "      <td>.</td>\n",
       "      <td>gene_id \"ENSMUSG00000102693.2\"; gene_type \"TEC...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>chr1</td>\n",
       "      <td>HAVANA</td>\n",
       "      <td>transcript</td>\n",
       "      <td>3143476</td>\n",
       "      <td>3144545</td>\n",
       "      <td>.</td>\n",
       "      <td>+</td>\n",
       "      <td>.</td>\n",
       "      <td>gene_id \"ENSMUSG00000102693.2\"; transcript_id ...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>chr1</td>\n",
       "      <td>HAVANA</td>\n",
       "      <td>exon</td>\n",
       "      <td>3143476</td>\n",
       "      <td>3144545</td>\n",
       "      <td>.</td>\n",
       "      <td>+</td>\n",
       "      <td>.</td>\n",
       "      <td>gene_id \"ENSMUSG00000102693.2\"; transcript_id ...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>chr1</td>\n",
       "      <td>ENSEMBL</td>\n",
       "      <td>gene</td>\n",
       "      <td>3172239</td>\n",
       "      <td>3172348</td>\n",
       "      <td>.</td>\n",
       "      <td>+</td>\n",
       "      <td>.</td>\n",
       "      <td>gene_id \"ENSMUSG00000064842.3\"; gene_type \"snR...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>chr1</td>\n",
       "      <td>ENSEMBL</td>\n",
       "      <td>transcript</td>\n",
       "      <td>3172239</td>\n",
       "      <td>3172348</td>\n",
       "      <td>.</td>\n",
       "      <td>+</td>\n",
       "      <td>.</td>\n",
       "      <td>gene_id \"ENSMUSG00000064842.3\"; transcript_id ...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>...</th>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1903937</th>\n",
       "      <td>chrM</td>\n",
       "      <td>ENSEMBL</td>\n",
       "      <td>transcript</td>\n",
       "      <td>15289</td>\n",
       "      <td>15355</td>\n",
       "      <td>.</td>\n",
       "      <td>+</td>\n",
       "      <td>.</td>\n",
       "      <td>gene_id \"ENSMUSG00000064371.1\"; transcript_id ...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1903938</th>\n",
       "      <td>chrM</td>\n",
       "      <td>ENSEMBL</td>\n",
       "      <td>exon</td>\n",
       "      <td>15289</td>\n",
       "      <td>15355</td>\n",
       "      <td>.</td>\n",
       "      <td>+</td>\n",
       "      <td>.</td>\n",
       "      <td>gene_id \"ENSMUSG00000064371.1\"; transcript_id ...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1903939</th>\n",
       "      <td>chrM</td>\n",
       "      <td>ENSEMBL</td>\n",
       "      <td>gene</td>\n",
       "      <td>15356</td>\n",
       "      <td>15422</td>\n",
       "      <td>.</td>\n",
       "      <td>-</td>\n",
       "      <td>.</td>\n",
       "      <td>gene_id \"ENSMUSG00000064372.1\"; gene_type \"Mt_...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1903940</th>\n",
       "      <td>chrM</td>\n",
       "      <td>ENSEMBL</td>\n",
       "      <td>transcript</td>\n",
       "      <td>15356</td>\n",
       "      <td>15422</td>\n",
       "      <td>.</td>\n",
       "      <td>-</td>\n",
       "      <td>.</td>\n",
       "      <td>gene_id \"ENSMUSG00000064372.1\"; transcript_id ...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1903941</th>\n",
       "      <td>chrM</td>\n",
       "      <td>ENSEMBL</td>\n",
       "      <td>exon</td>\n",
       "      <td>15356</td>\n",
       "      <td>15422</td>\n",
       "      <td>.</td>\n",
       "      <td>-</td>\n",
       "      <td>.</td>\n",
       "      <td>gene_id \"ENSMUSG00000064372.1\"; transcript_id ...</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>1903942 rows × 9 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "            0        1           2        3        4  5  6  7  \\\n",
       "0        chr1   HAVANA        gene  3143476  3144545  .  +  .   \n",
       "1        chr1   HAVANA  transcript  3143476  3144545  .  +  .   \n",
       "2        chr1   HAVANA        exon  3143476  3144545  .  +  .   \n",
       "3        chr1  ENSEMBL        gene  3172239  3172348  .  +  .   \n",
       "4        chr1  ENSEMBL  transcript  3172239  3172348  .  +  .   \n",
       "...       ...      ...         ...      ...      ... .. .. ..   \n",
       "1903937  chrM  ENSEMBL  transcript    15289    15355  .  +  .   \n",
       "1903938  chrM  ENSEMBL        exon    15289    15355  .  +  .   \n",
       "1903939  chrM  ENSEMBL        gene    15356    15422  .  -  .   \n",
       "1903940  chrM  ENSEMBL  transcript    15356    15422  .  -  .   \n",
       "1903941  chrM  ENSEMBL        exon    15356    15422  .  -  .   \n",
       "\n",
       "                                                         8  \n",
       "0        gene_id \"ENSMUSG00000102693.2\"; gene_type \"TEC...  \n",
       "1        gene_id \"ENSMUSG00000102693.2\"; transcript_id ...  \n",
       "2        gene_id \"ENSMUSG00000102693.2\"; transcript_id ...  \n",
       "3        gene_id \"ENSMUSG00000064842.3\"; gene_type \"snR...  \n",
       "4        gene_id \"ENSMUSG00000064842.3\"; transcript_id ...  \n",
       "...                                                    ...  \n",
       "1903937  gene_id \"ENSMUSG00000064371.1\"; transcript_id ...  \n",
       "1903938  gene_id \"ENSMUSG00000064371.1\"; transcript_id ...  \n",
       "1903939  gene_id \"ENSMUSG00000064372.1\"; gene_type \"Mt_...  \n",
       "1903940  gene_id \"ENSMUSG00000064372.1\"; transcript_id ...  \n",
       "1903941  gene_id \"ENSMUSG00000064372.1\"; transcript_id ...  \n",
       "\n",
       "[1903942 rows x 9 columns]"
      ]
     },
     "execution_count": 2,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "f = '../data_dump/gencode.vM32.annotation.gtf.gz'\n",
    "df = pd.read_csv(f, sep=\"\\t\", skiprows=5, header=None)\n",
    "df"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 3,
   "id": "b34a1e78-ab9a-402d-913c-27c3f4f044ce",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>0</th>\n",
       "      <th>1</th>\n",
       "      <th>2</th>\n",
       "      <th>3</th>\n",
       "      <th>4</th>\n",
       "      <th>5</th>\n",
       "      <th>6</th>\n",
       "      <th>7</th>\n",
       "      <th>8</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>chr1</td>\n",
       "      <td>HAVANA</td>\n",
       "      <td>gene</td>\n",
       "      <td>3143476</td>\n",
       "      <td>3144545</td>\n",
       "      <td>.</td>\n",
       "      <td>+</td>\n",
       "      <td>.</td>\n",
       "      <td>gene_id \"ENSMUSG00000102693.2\"; gene_type \"TEC...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>chr1</td>\n",
       "      <td>ENSEMBL</td>\n",
       "      <td>gene</td>\n",
       "      <td>3172239</td>\n",
       "      <td>3172348</td>\n",
       "      <td>.</td>\n",
       "      <td>+</td>\n",
       "      <td>.</td>\n",
       "      <td>gene_id \"ENSMUSG00000064842.3\"; gene_type \"snR...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>6</th>\n",
       "      <td>chr1</td>\n",
       "      <td>HAVANA</td>\n",
       "      <td>gene</td>\n",
       "      <td>3276124</td>\n",
       "      <td>3741721</td>\n",
       "      <td>.</td>\n",
       "      <td>-</td>\n",
       "      <td>.</td>\n",
       "      <td>gene_id \"ENSMUSG00000051951.6\"; gene_type \"pro...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>24</th>\n",
       "      <td>chr1</td>\n",
       "      <td>HAVANA</td>\n",
       "      <td>gene</td>\n",
       "      <td>3322980</td>\n",
       "      <td>3323459</td>\n",
       "      <td>.</td>\n",
       "      <td>+</td>\n",
       "      <td>.</td>\n",
       "      <td>gene_id \"ENSMUSG00000102851.2\"; gene_type \"pro...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>27</th>\n",
       "      <td>chr1</td>\n",
       "      <td>HAVANA</td>\n",
       "      <td>gene</td>\n",
       "      <td>3435954</td>\n",
       "      <td>3438772</td>\n",
       "      <td>.</td>\n",
       "      <td>-</td>\n",
       "      <td>.</td>\n",
       "      <td>gene_id \"ENSMUSG00000103377.2\"; gene_type \"TEC...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>...</th>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1903921</th>\n",
       "      <td>chrM</td>\n",
       "      <td>ENSEMBL</td>\n",
       "      <td>gene</td>\n",
       "      <td>13552</td>\n",
       "      <td>14070</td>\n",
       "      <td>.</td>\n",
       "      <td>-</td>\n",
       "      <td>.</td>\n",
       "      <td>gene_id \"ENSMUSG00000064368.1\"; gene_type \"pro...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1903928</th>\n",
       "      <td>chrM</td>\n",
       "      <td>ENSEMBL</td>\n",
       "      <td>gene</td>\n",
       "      <td>14071</td>\n",
       "      <td>14139</td>\n",
       "      <td>.</td>\n",
       "      <td>-</td>\n",
       "      <td>.</td>\n",
       "      <td>gene_id \"ENSMUSG00000064369.1\"; gene_type \"Mt_...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1903931</th>\n",
       "      <td>chrM</td>\n",
       "      <td>ENSEMBL</td>\n",
       "      <td>gene</td>\n",
       "      <td>14145</td>\n",
       "      <td>15288</td>\n",
       "      <td>.</td>\n",
       "      <td>+</td>\n",
       "      <td>.</td>\n",
       "      <td>gene_id \"ENSMUSG00000064370.1\"; gene_type \"pro...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1903936</th>\n",
       "      <td>chrM</td>\n",
       "      <td>ENSEMBL</td>\n",
       "      <td>gene</td>\n",
       "      <td>15289</td>\n",
       "      <td>15355</td>\n",
       "      <td>.</td>\n",
       "      <td>+</td>\n",
       "      <td>.</td>\n",
       "      <td>gene_id \"ENSMUSG00000064371.1\"; gene_type \"Mt_...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1903939</th>\n",
       "      <td>chrM</td>\n",
       "      <td>ENSEMBL</td>\n",
       "      <td>gene</td>\n",
       "      <td>15356</td>\n",
       "      <td>15422</td>\n",
       "      <td>.</td>\n",
       "      <td>-</td>\n",
       "      <td>.</td>\n",
       "      <td>gene_id \"ENSMUSG00000064372.1\"; gene_type \"Mt_...</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>56953 rows × 9 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "            0        1     2        3        4  5  6  7  \\\n",
       "0        chr1   HAVANA  gene  3143476  3144545  .  +  .   \n",
       "3        chr1  ENSEMBL  gene  3172239  3172348  .  +  .   \n",
       "6        chr1   HAVANA  gene  3276124  3741721  .  -  .   \n",
       "24       chr1   HAVANA  gene  3322980  3323459  .  +  .   \n",
       "27       chr1   HAVANA  gene  3435954  3438772  .  -  .   \n",
       "...       ...      ...   ...      ...      ... .. .. ..   \n",
       "1903921  chrM  ENSEMBL  gene    13552    14070  .  -  .   \n",
       "1903928  chrM  ENSEMBL  gene    14071    14139  .  -  .   \n",
       "1903931  chrM  ENSEMBL  gene    14145    15288  .  +  .   \n",
       "1903936  chrM  ENSEMBL  gene    15289    15355  .  +  .   \n",
       "1903939  chrM  ENSEMBL  gene    15356    15422  .  -  .   \n",
       "\n",
       "                                                         8  \n",
       "0        gene_id \"ENSMUSG00000102693.2\"; gene_type \"TEC...  \n",
       "3        gene_id \"ENSMUSG00000064842.3\"; gene_type \"snR...  \n",
       "6        gene_id \"ENSMUSG00000051951.6\"; gene_type \"pro...  \n",
       "24       gene_id \"ENSMUSG00000102851.2\"; gene_type \"pro...  \n",
       "27       gene_id \"ENSMUSG00000103377.2\"; gene_type \"TEC...  \n",
       "...                                                    ...  \n",
       "1903921  gene_id \"ENSMUSG00000064368.1\"; gene_type \"pro...  \n",
       "1903928  gene_id \"ENSMUSG00000064369.1\"; gene_type \"Mt_...  \n",
       "1903931  gene_id \"ENSMUSG00000064370.1\"; gene_type \"pro...  \n",
       "1903936  gene_id \"ENSMUSG00000064371.1\"; gene_type \"Mt_...  \n",
       "1903939  gene_id \"ENSMUSG00000064372.1\"; gene_type \"Mt_...  \n",
       "\n",
       "[56953 rows x 9 columns]"
      ]
     },
     "execution_count": 3,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "dfg = df[df[2] == 'gene'].copy()\n",
    "dfg "
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 4,
   "id": "be52685e-ebfe-451f-838f-db5b2a2e91c5",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>0</th>\n",
       "      <th>1</th>\n",
       "      <th>2</th>\n",
       "      <th>3</th>\n",
       "      <th>4</th>\n",
       "      <th>5</th>\n",
       "      <th>6</th>\n",
       "      <th>7</th>\n",
       "      <th>8</th>\n",
       "      <th>gene_id</th>\n",
       "      <th>gene_name</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>chr1</td>\n",
       "      <td>HAVANA</td>\n",
       "      <td>gene</td>\n",
       "      <td>3143476</td>\n",
       "      <td>3144545</td>\n",
       "      <td>.</td>\n",
       "      <td>+</td>\n",
       "      <td>.</td>\n",
       "      <td>gene_id \"ENSMUSG00000102693.2\"; gene_type \"TEC...</td>\n",
       "      <td>ENSMUSG00000102693.2</td>\n",
       "      <td>4933401J01Rik</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>chr1</td>\n",
       "      <td>ENSEMBL</td>\n",
       "      <td>gene</td>\n",
       "      <td>3172239</td>\n",
       "      <td>3172348</td>\n",
       "      <td>.</td>\n",
       "      <td>+</td>\n",
       "      <td>.</td>\n",
       "      <td>gene_id \"ENSMUSG00000064842.3\"; gene_type \"snR...</td>\n",
       "      <td>ENSMUSG00000064842.3</td>\n",
       "      <td>Gm26206</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>6</th>\n",
       "      <td>chr1</td>\n",
       "      <td>HAVANA</td>\n",
       "      <td>gene</td>\n",
       "      <td>3276124</td>\n",
       "      <td>3741721</td>\n",
       "      <td>.</td>\n",
       "      <td>-</td>\n",
       "      <td>.</td>\n",
       "      <td>gene_id \"ENSMUSG00000051951.6\"; gene_type \"pro...</td>\n",
       "      <td>ENSMUSG00000051951.6</td>\n",
       "      <td>Xkr4</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>24</th>\n",
       "      <td>chr1</td>\n",
       "      <td>HAVANA</td>\n",
       "      <td>gene</td>\n",
       "      <td>3322980</td>\n",
       "      <td>3323459</td>\n",
       "      <td>.</td>\n",
       "      <td>+</td>\n",
       "      <td>.</td>\n",
       "      <td>gene_id \"ENSMUSG00000102851.2\"; gene_type \"pro...</td>\n",
       "      <td>ENSMUSG00000102851.2</td>\n",
       "      <td>Gm18956</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>27</th>\n",
       "      <td>chr1</td>\n",
       "      <td>HAVANA</td>\n",
       "      <td>gene</td>\n",
       "      <td>3435954</td>\n",
       "      <td>3438772</td>\n",
       "      <td>.</td>\n",
       "      <td>-</td>\n",
       "      <td>.</td>\n",
       "      <td>gene_id \"ENSMUSG00000103377.2\"; gene_type \"TEC...</td>\n",
       "      <td>ENSMUSG00000103377.2</td>\n",
       "      <td>Gm37180</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>...</th>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1903921</th>\n",
       "      <td>chrM</td>\n",
       "      <td>ENSEMBL</td>\n",
       "      <td>gene</td>\n",
       "      <td>13552</td>\n",
       "      <td>14070</td>\n",
       "      <td>.</td>\n",
       "      <td>-</td>\n",
       "      <td>.</td>\n",
       "      <td>gene_id \"ENSMUSG00000064368.1\"; gene_type \"pro...</td>\n",
       "      <td>ENSMUSG00000064368.1</td>\n",
       "      <td>mt-Nd6</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1903928</th>\n",
       "      <td>chrM</td>\n",
       "      <td>ENSEMBL</td>\n",
       "      <td>gene</td>\n",
       "      <td>14071</td>\n",
       "      <td>14139</td>\n",
       "      <td>.</td>\n",
       "      <td>-</td>\n",
       "      <td>.</td>\n",
       "      <td>gene_id \"ENSMUSG00000064369.1\"; gene_type \"Mt_...</td>\n",
       "      <td>ENSMUSG00000064369.1</td>\n",
       "      <td>mt-Te</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1903931</th>\n",
       "      <td>chrM</td>\n",
       "      <td>ENSEMBL</td>\n",
       "      <td>gene</td>\n",
       "      <td>14145</td>\n",
       "      <td>15288</td>\n",
       "      <td>.</td>\n",
       "      <td>+</td>\n",
       "      <td>.</td>\n",
       "      <td>gene_id \"ENSMUSG00000064370.1\"; gene_type \"pro...</td>\n",
       "      <td>ENSMUSG00000064370.1</td>\n",
       "      <td>mt-Cytb</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1903936</th>\n",
       "      <td>chrM</td>\n",
       "      <td>ENSEMBL</td>\n",
       "      <td>gene</td>\n",
       "      <td>15289</td>\n",
       "      <td>15355</td>\n",
       "      <td>.</td>\n",
       "      <td>+</td>\n",
       "      <td>.</td>\n",
       "      <td>gene_id \"ENSMUSG00000064371.1\"; gene_type \"Mt_...</td>\n",
       "      <td>ENSMUSG00000064371.1</td>\n",
       "      <td>mt-Tt</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1903939</th>\n",
       "      <td>chrM</td>\n",
       "      <td>ENSEMBL</td>\n",
       "      <td>gene</td>\n",
       "      <td>15356</td>\n",
       "      <td>15422</td>\n",
       "      <td>.</td>\n",
       "      <td>-</td>\n",
       "      <td>.</td>\n",
       "      <td>gene_id \"ENSMUSG00000064372.1\"; gene_type \"Mt_...</td>\n",
       "      <td>ENSMUSG00000064372.1</td>\n",
       "      <td>mt-Tp</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>56953 rows × 11 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "            0        1     2        3        4  5  6  7  \\\n",
       "0        chr1   HAVANA  gene  3143476  3144545  .  +  .   \n",
       "3        chr1  ENSEMBL  gene  3172239  3172348  .  +  .   \n",
       "6        chr1   HAVANA  gene  3276124  3741721  .  -  .   \n",
       "24       chr1   HAVANA  gene  3322980  3323459  .  +  .   \n",
       "27       chr1   HAVANA  gene  3435954  3438772  .  -  .   \n",
       "...       ...      ...   ...      ...      ... .. .. ..   \n",
       "1903921  chrM  ENSEMBL  gene    13552    14070  .  -  .   \n",
       "1903928  chrM  ENSEMBL  gene    14071    14139  .  -  .   \n",
       "1903931  chrM  ENSEMBL  gene    14145    15288  .  +  .   \n",
       "1903936  chrM  ENSEMBL  gene    15289    15355  .  +  .   \n",
       "1903939  chrM  ENSEMBL  gene    15356    15422  .  -  .   \n",
       "\n",
       "                                                         8  \\\n",
       "0        gene_id \"ENSMUSG00000102693.2\"; gene_type \"TEC...   \n",
       "3        gene_id \"ENSMUSG00000064842.3\"; gene_type \"snR...   \n",
       "6        gene_id \"ENSMUSG00000051951.6\"; gene_type \"pro...   \n",
       "24       gene_id \"ENSMUSG00000102851.2\"; gene_type \"pro...   \n",
       "27       gene_id \"ENSMUSG00000103377.2\"; gene_type \"TEC...   \n",
       "...                                                    ...   \n",
       "1903921  gene_id \"ENSMUSG00000064368.1\"; gene_type \"pro...   \n",
       "1903928  gene_id \"ENSMUSG00000064369.1\"; gene_type \"Mt_...   \n",
       "1903931  gene_id \"ENSMUSG00000064370.1\"; gene_type \"pro...   \n",
       "1903936  gene_id \"ENSMUSG00000064371.1\"; gene_type \"Mt_...   \n",
       "1903939  gene_id \"ENSMUSG00000064372.1\"; gene_type \"Mt_...   \n",
       "\n",
       "                      gene_id      gene_name  \n",
       "0        ENSMUSG00000102693.2  4933401J01Rik  \n",
       "3        ENSMUSG00000064842.3        Gm26206  \n",
       "6        ENSMUSG00000051951.6           Xkr4  \n",
       "24       ENSMUSG00000102851.2        Gm18956  \n",
       "27       ENSMUSG00000103377.2        Gm37180  \n",
       "...                       ...            ...  \n",
       "1903921  ENSMUSG00000064368.1         mt-Nd6  \n",
       "1903928  ENSMUSG00000064369.1          mt-Te  \n",
       "1903931  ENSMUSG00000064370.1        mt-Cytb  \n",
       "1903936  ENSMUSG00000064371.1          mt-Tt  \n",
       "1903939  ENSMUSG00000064372.1          mt-Tp  \n",
       "\n",
       "[56953 rows x 11 columns]"
      ]
     },
     "execution_count": 4,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "dfg['gene_id']   = dfg[8].apply(lambda x: x.split(';')[0].split(' ')[1].strip('\\\"')) \n",
    "dfg['gene_name'] = dfg[8].apply(lambda x: x.split(';')[2].split(' ')[2].strip('\\\"'))\n",
    "dfg"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 10,
   "id": "82850156-3ce5-44d3-9b93-1afc732ce33e",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>0</th>\n",
       "      <th>3</th>\n",
       "      <th>4</th>\n",
       "      <th>6</th>\n",
       "      <th>gene_id</th>\n",
       "      <th>gene_name</th>\n",
       "      <th>gid_short</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>chr1</td>\n",
       "      <td>3143476</td>\n",
       "      <td>3144545</td>\n",
       "      <td>+</td>\n",
       "      <td>ENSMUSG00000102693.2</td>\n",
       "      <td>4933401J01Rik</td>\n",
       "      <td>ENSMUSG00000102693</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>chr1</td>\n",
       "      <td>3172239</td>\n",
       "      <td>3172348</td>\n",
       "      <td>+</td>\n",
       "      <td>ENSMUSG00000064842.3</td>\n",
       "      <td>Gm26206</td>\n",
       "      <td>ENSMUSG00000064842</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>6</th>\n",
       "      <td>chr1</td>\n",
       "      <td>3276124</td>\n",
       "      <td>3741721</td>\n",
       "      <td>-</td>\n",
       "      <td>ENSMUSG00000051951.6</td>\n",
       "      <td>Xkr4</td>\n",
       "      <td>ENSMUSG00000051951</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>24</th>\n",
       "      <td>chr1</td>\n",
       "      <td>3322980</td>\n",
       "      <td>3323459</td>\n",
       "      <td>+</td>\n",
       "      <td>ENSMUSG00000102851.2</td>\n",
       "      <td>Gm18956</td>\n",
       "      <td>ENSMUSG00000102851</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>27</th>\n",
       "      <td>chr1</td>\n",
       "      <td>3435954</td>\n",
       "      <td>3438772</td>\n",
       "      <td>-</td>\n",
       "      <td>ENSMUSG00000103377.2</td>\n",
       "      <td>Gm37180</td>\n",
       "      <td>ENSMUSG00000103377</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>...</th>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1903921</th>\n",
       "      <td>chrM</td>\n",
       "      <td>13552</td>\n",
       "      <td>14070</td>\n",
       "      <td>-</td>\n",
       "      <td>ENSMUSG00000064368.1</td>\n",
       "      <td>mt-Nd6</td>\n",
       "      <td>ENSMUSG00000064368</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1903928</th>\n",
       "      <td>chrM</td>\n",
       "      <td>14071</td>\n",
       "      <td>14139</td>\n",
       "      <td>-</td>\n",
       "      <td>ENSMUSG00000064369.1</td>\n",
       "      <td>mt-Te</td>\n",
       "      <td>ENSMUSG00000064369</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1903931</th>\n",
       "      <td>chrM</td>\n",
       "      <td>14145</td>\n",
       "      <td>15288</td>\n",
       "      <td>+</td>\n",
       "      <td>ENSMUSG00000064370.1</td>\n",
       "      <td>mt-Cytb</td>\n",
       "      <td>ENSMUSG00000064370</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1903936</th>\n",
       "      <td>chrM</td>\n",
       "      <td>15289</td>\n",
       "      <td>15355</td>\n",
       "      <td>+</td>\n",
       "      <td>ENSMUSG00000064371.1</td>\n",
       "      <td>mt-Tt</td>\n",
       "      <td>ENSMUSG00000064371</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1903939</th>\n",
       "      <td>chrM</td>\n",
       "      <td>15356</td>\n",
       "      <td>15422</td>\n",
       "      <td>-</td>\n",
       "      <td>ENSMUSG00000064372.1</td>\n",
       "      <td>mt-Tp</td>\n",
       "      <td>ENSMUSG00000064372</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>56953 rows × 7 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "            0        3        4  6               gene_id      gene_name  \\\n",
       "0        chr1  3143476  3144545  +  ENSMUSG00000102693.2  4933401J01Rik   \n",
       "3        chr1  3172239  3172348  +  ENSMUSG00000064842.3        Gm26206   \n",
       "6        chr1  3276124  3741721  -  ENSMUSG00000051951.6           Xkr4   \n",
       "24       chr1  3322980  3323459  +  ENSMUSG00000102851.2        Gm18956   \n",
       "27       chr1  3435954  3438772  -  ENSMUSG00000103377.2        Gm37180   \n",
       "...       ...      ...      ... ..                   ...            ...   \n",
       "1903921  chrM    13552    14070  -  ENSMUSG00000064368.1         mt-Nd6   \n",
       "1903928  chrM    14071    14139  -  ENSMUSG00000064369.1          mt-Te   \n",
       "1903931  chrM    14145    15288  +  ENSMUSG00000064370.1        mt-Cytb   \n",
       "1903936  chrM    15289    15355  +  ENSMUSG00000064371.1          mt-Tt   \n",
       "1903939  chrM    15356    15422  -  ENSMUSG00000064372.1          mt-Tp   \n",
       "\n",
       "                  gid_short  \n",
       "0        ENSMUSG00000102693  \n",
       "3        ENSMUSG00000064842  \n",
       "6        ENSMUSG00000051951  \n",
       "24       ENSMUSG00000102851  \n",
       "27       ENSMUSG00000103377  \n",
       "...                     ...  \n",
       "1903921  ENSMUSG00000064368  \n",
       "1903928  ENSMUSG00000064369  \n",
       "1903931  ENSMUSG00000064370  \n",
       "1903936  ENSMUSG00000064371  \n",
       "1903939  ENSMUSG00000064372  \n",
       "\n",
       "[56953 rows x 7 columns]"
      ]
     },
     "execution_count": 10,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "dfgbed = dfg[[0, 3, 4, 6, 'gene_id', 'gene_name',]].copy()\n",
    "dfgbed['gid_short'] = dfgbed['gene_id'].apply(lambda x: x.split('.')[0])\n",
    "dfgbed"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 11,
   "id": "926820e0-156a-4877-a0c8-c227aad8e78a",
   "metadata": {},
   "outputs": [],
   "source": [
    "fout = '../data_dump/gencode.vM32.allgenes.bed'\n",
    "dfgbed.to_csv(fout, sep=\"\\t\", header=False, index=False)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 12,
   "id": "a5fa0c2e-053b-4fed-8148-4f0e2932296d",
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "chr1\t3143476\t3144545\t+\tENSMUSG00000102693.2\t4933401J01Rik\tENSMUSG00000102693\n",
      "chr1\t3172239\t3172348\t+\tENSMUSG00000064842.3\tGm26206\tENSMUSG00000064842\n",
      "chr1\t3276124\t3741721\t-\tENSMUSG00000051951.6\tXkr4\tENSMUSG00000051951\n",
      "chr1\t3322980\t3323459\t+\tENSMUSG00000102851.2\tGm18956\tENSMUSG00000102851\n",
      "chr1\t3435954\t3438772\t-\tENSMUSG00000103377.2\tGm37180\tENSMUSG00000103377\n",
      "chr1\t3445779\t3448011\t-\tENSMUSG00000104017.2\tGm37363\tENSMUSG00000104017\n",
      "chr1\t3535200\t3537508\t-\tENSMUSG00000103025.2\tGm37686\tENSMUSG00000103025\n",
      "chr1\t3536810\t3583776\t+\tENSMUSG00000089699.2\tGm1992\tENSMUSG00000089699\n",
      "chr1\t3582674\t3584730\t-\tENSMUSG00000103201.2\tGm37329\tENSMUSG00000103201\n",
      "chr1\t3602018\t3602943\t+\tENSMUSG00000103147.2\tGm7341\tENSMUSG00000103147\n"
     ]
    }
   ],
   "source": [
    "!head $fout"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "567eefdc-d76e-46d1-bc00-c9b0970cdf9e",
   "metadata": {},
   "outputs": [],
   "source": []
  }
 ],
 "metadata": {
  "kernelspec": {
   "display_name": "Python 3 (ipykernel)",
   "language": "python",
   "name": "python3"
  },
  "language_info": {
   "codemirror_mode": {
    "name": "ipython",
    "version": 3
   },
   "file_extension": ".py",
   "mimetype": "text/x-python",
   "name": "python",
   "nbconvert_exporter": "python",
   "pygments_lexer": "ipython3",
   "version": "3.10.2"
  }
 },
 "nbformat": 4,
 "nbformat_minor": 5
}
